### PR TITLE
Python: Limit set of globals that may be built-ins

### PR DESCRIPTION
**Pull request by tausbn**
I am very tempted to leave out the constants, or at the very least
`False`, `True`, and `None`, as these have _many_ occurrences in the
average codebase, and are not terribly useful at the API-graph level.

If we really do want to capture "nodes that refer to such and such
constant", then I think a better solution would be to create classes
extending `DataFlow::Node` to facilitate this.

---
No change note required -- this only gets rid of junk that should never have been there in the first place.

As an added bonus, I've also ported the `py/use-of-input` query to use API graphs.

I'm not entirely convinced that this way of handling shadowing of built-ins is the best possible. One might make the argument, for instance, that if something redefines `print`, say, then it's likely a function that acts as a drop-in replacement for `print` (and therefore also of interest when looking for instances of the `print` built-in). With that in mind, it may be better to keep the redefined built-ins, and perhaps allow for some other way of checking if they may have been redefined.

### CODE REVIEW

File: python/ql/src/semmle/python/ApiGraphs.qll
Comment (RasmusWL) [inline python/ql/src/semmle/python/ApiGraphs.qll:352]:
```suggestion
```

As I see it, we're not using this import anymore, and I don't think there are any side-effects from having the classes from this module in scope that we care about?

**Review by RasmusWL — COMMENTED** [review]
I had to remind myself why this was a problem (sorry for the long delay in reviewing)

> Oof. It seems our definition of `API::builtin` is a bit too permissive. We get things like the `random` module (because apparently that's in `Builtin`) and thus erroneously think of references to `random` (even when imported correctly) as references to `builtins.random` which is just nonsense. 

### Python 2 vs Python 3

I'm not sure I understand whether we gain anything from only allowing Python 2 specific builtins when we extracted as Python 2 (and vice versa). If the example below was extracted as Python 3, it would seems a bit strange to me that `API::builtin("xrange").getACall()` would not have any results.

I'm tempted to suggest we just allow all builtins no matter what version they come from, removing the `major_version() = ...` checks, but I'd like to hear whether there's a good reason _not_ to do that :smile: 

```py
import sys
PY2 = sys.version_info[0] == 2

if PY2:
    ids = xrange(100)
else:
    ids = range(100)

for x in ids:
    print(x)
```

### To handle overwrites or not

I have a hard time judging whether the change below makes this better or not. I could probably come up with examples where either version is preferable. I don't really know, so not going to make any complaints about this :sweat_smile: Do you have plans for altering this in the future to be more based around data-flow? (for example, considering a fake `from builtins import *` in the top of every file)

```diff
-possible_builtin_accessed_in_module(n, name, m)
+possible_builtin_accessed_in_module(n, name, m) and
+not possible_builtin_defined_in_module(name, m)
```

File: python/ql/src/semmle/python/ApiGraphs.qll
Comment (tausbn) [reply python/ql/src/semmle/python/ApiGraphs.qll:352]: Good catch. Indeed, there is no reason to keep this around.

File: python/ql/src/semmle/python/ApiGraphs.qll
Comment (RasmusWL) [reply python/ql/src/semmle/python/ApiGraphs.qll:364]: Thanks for explaining :pray: sounds very reasonable to me 

> I should probably add a comment to this effect, in case we wish to redo this in the future.

That sounds very nice :+1:

**Comment by tausbn** [comment]
You bring up a fair point about just recognising all built-ins always, regardless of what version of Python we're extracting the code as. I will remove the `major_version` conjuncts (but keep the set literals separate for now, in case it becomes relevant in the future).

As for the overwrites, I imagine we'll want to extend the analysis in the future to handle things like `import *`. My main reason for opting for the present version is that if we don't, something like
```py
def input(s):
    return raw_input(s)
```
would result in false positives for the `py/use-of-input` query (as any call to `input` would be seen as a direct use of the `input` built-in, ignoring the fact that it has been redefined). I guess one could argue, though, that this logic could be moved into that particular query, and keep the more permissive "anything that looks like a built-in" logic we had previously. I'm fine with either solution. I'm also happy to just say "ship it" for now, and figure out if this becomes an issue.

**Comment by RasmusWL** [comment]
With the above :point_up: context in mind, I think it sounds the handling of built-in overrides sounds introduced in this PR sounds reasonable :+1: (as long as we remember to further improve this in the future :sweat_smile:)

File: python/ql/src/semmle/python/ApiGraphs.qll
Comment (RasmusWL) [inline python/ql/src/semmle/python/ApiGraphs.qll:355]:
```suggestion
      // These lists were created by inspecting the `builtins` and `__builtin__` modules in
      // Python 3 and 2 respectively, using the `dir` built-in.
```

**Review by RasmusWL — DISMISSED** [review]
Besides the little nitpick about which version has `builtins` and which version has `__builtin__`, LGTM :+1:

**Review by RasmusWL — APPROVED** [review]

**Pull request merged**
